File: SuperGrid/VB/GridStateEditControl.vb
Imports System.ComponentModel
Imports DevComponents.DotNetBar.SuperGrid
Imports DevComponents.DotNetBar.SuperGrid.Style

Namespace SuperGridDemo
	'''<summary>
	''' GridStateEditControl
	'''</summary>
	<ToolboxItem(False)> _
	Public Class GridStateEditControl
		Inherits StateControl
        Implements IGridCellEditControl

#Region "Private variables"

        Private _Cell As GridCell
        Private _EditorPanel As EditorPanel
        Private _EditorCellBitmap As Bitmap

        Private _ValueChanged As Boolean
        Private _SuspendUpdate As Boolean

        Private _StretchBehavior As StretchBehavior = StretchBehavior.None

#End Region

        '''<summary>
        ''' GridStateEditControl
        '''</summary>
        Public Sub New()
            AddHandler StateChanged, AddressOf ControlStateChanged
        End Sub

#Region "ControlStateChanged"

        ''' <summary>
        ''' Routine called when the control 'state' has changed.
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        Private Sub ControlStateChanged(ByVal sender As Object, ByVal e As StateChangedEventArgs)
            If _SuspendUpdate = False Then
                _Cell.Value = e.NewValue

                _Cell.EditorValueChanged(Me)
            End If
        End Sub

#End Region

#Region "OnInvalidated"

        Protected Overrides Sub OnInvalidated(ByVal e As InvalidateEventArgs)
            MyBase.OnInvalidated(e)

            If _Cell IsNot Nothing AndAlso _SuspendUpdate = False Then
                _Cell.InvalidateRender()
            End If
        End Sub

#End Region

#Region "OnPaint"

        Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
            _Cell.PaintEditorBackground(e, Me)

            MyBase.OnPaint(e)
        End Sub

#End Region

#Region "GetValue"

        '''<summary>
        ''' GetValue
        '''</summary>
        '''<param name="value"></param>
        '''<returns></returns>
        '''<exception cref="Exception"></exception>
        Public Overridable Function GetValue(ByVal value As Object) As String
            Dim panel As GridPanel = _Cell.GridPanel

            If value Is Nothing OrElse (panel.NullValue = NullValue.DBNull AndAlso value Is DBNull.Value) Then
                Return ("")
            End If

            Return (Convert.ToString(value))
        End Function

#End Region

#Region "IGridCellEditControl Members"

#Region "Public properties"

#Region "CanInterrupt"

        Public ReadOnly Property CanInterrupt() As Boolean Implements IGridCellEditControl.CanInterrupt
            Get
                Return ((MouseButtons And MouseButtons.Left) <> MouseButtons.Left)
            End Get
        End Property

#End Region

#Region "CellEditMode"

        Public ReadOnly Property CellEditMode() As CellEditMode Implements IGridCellEditControl.CellEditMode
            Get
                Return (CellEditMode.InPlace)
            End Get
        End Property

#End Region

#Region "EditorCell"

        Public Property EditorCell() As GridCell Implements IGridCellEditControl.EditorCell
            Get
                Return (_Cell)
            End Get
            Set(ByVal value As GridCell)
                _Cell = value
            End Set
        End Property

#End Region

#Region "EditorCellBitmap"

        '''<summary>
        ''' EditorCellBitmap
        '''</summary>
        Public Property EditorCellBitmap() As Bitmap Implements IGridCellEditControl.EditorCellBitmap
            Get
                Return (_EditorCellBitmap)
            End Get

            Set(ByVal value As Bitmap)
                If _EditorCellBitmap IsNot Nothing Then
                    _EditorCellBitmap.Dispose()
                End If

                _EditorCellBitmap = value
            End Set
        End Property

#End Region

#Region "EditorFormattedValue"

        '''<summary>
        ''' EditorFormattedValue
        '''</summary>
        Public Overridable ReadOnly Property EditorFormattedValue() As String Implements IGridCellEditControl.EditorFormattedValue
            Get
                Return (Text)
            End Get
        End Property

#End Region

#Region "EditorPanel"

        Public Property EditorPanel() As EditorPanel Implements IGridCellEditControl.EditorPanel
            Get
                Return (_EditorPanel)
            End Get
            Set(ByVal value As EditorPanel)
                _EditorPanel = value
            End Set
        End Property

#End Region

#Region "EditorValue"

        Public Overridable Property EditorValue() As Object Implements IGridCellEditControl.EditorValue
            Get
                Return (Value)
            End Get
            Set(ByVal value As Object)
                value = GetValue(value)
            End Set
        End Property

#End Region

#Region "EditorValueChanged"

        Public Overridable Property EditorValueChanged() As Boolean Implements IGridCellEditControl.EditorValueChanged
            Get
                Return (_ValueChanged)
            End Get

            Set(ByVal value As Boolean)
                If _ValueChanged <> value Then
                    _ValueChanged = value

                    If value = True Then
                        _Cell.SetEditorDirty(Me)
                    End If
                End If
            End Set
        End Property

#End Region

#Region "EditorValueType"

        '''<summary>
        ''' EditorValueType
        '''</summary>
        Public Overridable ReadOnly Property EditorValueType() As Type Implements IGridCellEditControl.EditorValueType
            Get
                Return (GetType(String))
            End Get
        End Property

#End Region

#Region "StretchBehavior"

        Public Overridable ReadOnly Property StretchBehavior() As StretchBehavior Implements IGridCellEditControl.StretchBehavior
            Get
                Return (_StretchBehavior)
            End Get
        End Property

#End Region

#Region "SuspendUpdate"

        Public Property SuspendUpdate() As Boolean Implements IGridCellEditControl.SuspendUpdate
            Get
                Return (_SuspendUpdate)
            End Get
            Set(ByVal value As Boolean)
                _SuspendUpdate = value
            End Set
        End Property

#End Region

#Region "ValueChangeBehavior"

        Public Overridable ReadOnly Property ValueChangeBehavior() As ValueChangeBehavior Implements IGridCellEditControl.ValueChangeBehavior
            Get
                Return (ValueChangeBehavior.InvalidateLayout)
            End Get
        End Property

#End Region

#End Region

#Region "InitializeContext"

        '''<summary>
        ''' InitializeContext
        '''</summary>
        '''<param name="cell"></param>
        '''<param name="style"></param>
        Public Overridable Sub InitializeContext(ByVal cell As GridCell, ByVal style As CellVisualStyle) Implements IGridCellEditControl.InitializeContext
            _Cell = cell

            Value = GetValue(_Cell.Value)
            Enabled = (_Cell.IsReadOnly = False)

            _ValueChanged = False
        End Sub

#End Region

#Region "GetProposedSize"

        '''<summary>
        ''' GetProposedSize
        '''</summary>
        '''<param name="g"></param>
        '''<param name="cell"></param>
        '''<param name="style"></param>
        '''<param name="constraintSize"></param>
        '''<returns></returns>
        Public Overridable Function GetProposedSize(ByVal g As Graphics, ByVal cell As GridCell, ByVal style As CellVisualStyle, ByVal constraintSize As Size) As Size Implements IGridCellEditControl.GetProposedSize
            Return (GetPreferredSize(constraintSize))
        End Function

#End Region

#Region "Edit support"

#Region "BeginEdit"

        Public Overridable Function BeginEdit(ByVal selectAll As Boolean) As Boolean Implements IGridCellEditControl.BeginEdit
            Return (False)
        End Function

#End Region

#Region "EndEdit"

        Public Overridable Function EndEdit() As Boolean Implements IGridCellEditControl.EndEdit
            Return (False)
        End Function

#End Region

#Region "CancelEdit"

        Public Overridable Function CancelEdit() As Boolean Implements IGridCellEditControl.CancelEdit
            Return (False)
        End Function

#End Region

#End Region

#Region "CellRender"

        Public Overridable Sub CellRender(ByVal g As Graphics) Implements IGridCellEditControl.CellRender
            _Cell.CellRender(Me, g)
        End Sub

#End Region

#Region "Keyboard support"

#Region "CellKeyDown"

        '''<summary>
        ''' CellKeyDown
        '''</summary>
        Public Overridable Sub CellKeyDown(ByVal e As KeyEventArgs) Implements IGridCellEditControl.CellKeyDown
            ' If the user types a digit (0 - 9) then increment
            ' the state color by 1 at that state offset

            If e.KeyData >= Keys.D0 AndAlso e.KeyData <= Keys.D9 Then
                Dim index As Integer = e.KeyData - Keys.D0
                Dim value As String = DirectCast(EditorCell.Value, String)
                Dim s As String() = value.Split(","c)

                Dim k As Integer = Math.Max(index + 1, s.Length)
                Dim n As Integer() = New Integer(k - 1) {}

                For i As Integer = 0 To k - 1
                    If i < s.Length Then
                        n(i) = Val(s(i)(0))
                    Else
                        n(i) = 0
                    End If
                Next

                n(index) = (n(index) + 1) Mod 5

                Dim sb As System.Text.StringBuilder = New System.Text.StringBuilder()

                For i As Integer = 0 To k - 1
                    sb.Append(n(i))
                    sb.Append(",")
                Next

                sb.Length -= 1

                EditorCell.Value = sb.ToString()
            Else
                OnKeyDown(e)
            End If
        End Sub

#End Region

#Region "WantsInputKey"

        Public Overridable Function WantsInputKey(ByVal key As Keys, ByVal gridWantsKey As Boolean) As Boolean Implements IGridCellEditControl.WantsInputKey
            ' Lets handle all the digit keys

            If key >= Keys.D0 AndAlso key <= Keys.D9 Then
                Return (True)
            End If

            ' And let the grid handle the rest

            Return (gridWantsKey = False)
        End Function


#End Region

#End Region

#Region "Mouse support"

#Region "OnCellMouseMove"

        '''<summary>
        ''' OnCellMouseMove
        '''</summary>
        '''<param name="e"></param>
        Public Overridable Sub OnCellMouseMove(ByVal e As MouseEventArgs) Implements IGridCellEditControl.OnCellMouseMove
            OnMouseMove(e)
        End Sub

#End Region

#Region "OnCellMouseEnter"

        '''<summary>
        ''' OnCellMouseEnter
        '''</summary>
        '''<param name="e"></param>
        Public Overridable Sub OnCellMouseEnter(ByVal e As EventArgs) Implements IGridCellEditControl.OnCellMouseEnter
            OnMouseEnter(e)
        End Sub

#End Region

#Region "OnCellMouseLeave"

        '''<summary>
        ''' OnCellMouseLeave
        '''</summary>
        '''<param name="e"></param>
        Public Overridable Sub OnCellMouseLeave(ByVal e As EventArgs) Implements IGridCellEditControl.OnCellMouseLeave
            OnMouseLeave(e)
        End Sub

#End Region

#Region "OnCellMouseUp"

        '''<summary>
        ''' OnCellMouseUp
        '''</summary>
        '''<param name="e"></param>
        Public Overridable Sub OnCellMouseUp(ByVal e As MouseEventArgs) Implements IGridCellEditControl.OnCellMouseUp
            OnMouseUp(e)
        End Sub

#End Region

#Region "OnCellMouseDown"

        '''<summary>
        ''' OnCellMouseDown
        '''</summary>
        '''<param name="e"></param>
        Public Overridable Sub OnCellMouseDown(ByVal e As MouseEventArgs) Implements IGridCellEditControl.OnCellMouseDown
            OnMouseDown(e)
        End Sub

#End Region

#End Region

#End Region

    End Class
End Namespace

File: SuperGrid/C#/GridStateEditControl.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevComponents.DotNetBar.SuperGrid;
using DevComponents.DotNetBar.SuperGrid.Style;

namespace SuperGridDemo
{
    ///<summary>
    /// GridStateEditControl
    ///</summary>
    [ToolboxItem(false)]
    public class GridStateEditControl : StateControl, IGridCellEditControl
    {
        #region Private variables

        private GridCell _Cell;
        private EditorPanel _EditorPanel;
        private Bitmap _EditorCellBitmap;

        private bool _ValueChanged;
        private bool _SuspendUpdate;

        private StretchBehavior _StretchBehavior = StretchBehavior.None;

        #endregion

        ///<summary>
        /// GridStateEditControl
        ///</summary>
        public GridStateEditControl()
        {
            StateChanged += ControlStateChanged;
        }

        #region ControlStateChanged

        /// <summary>
        /// Routine called when the control 'state' has changed.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void ControlStateChanged(object sender, StateChangedEventArgs e)
        {
            if (_SuspendUpdate == false)
            {
                _Cell.Value = e.NewValue;

                _Cell.EditorValueChanged(this);
            }
        }

        #endregion

        #region OnInvalidated

        protected override void OnInvalidated(InvalidateEventArgs e)
        {
            base.OnInvalidated(e);

            if (_Cell != null && _SuspendUpdate == false)
                _Cell.InvalidateRender();
        }

        #endregion

        #region OnPaint

        protected override void OnPaint(PaintEventArgs e)
        {
            _Cell.PaintEditorBackground(e, this);

            base.OnPaint(e);
        }

        #endregion

        #region GetValue

        ///<summary>
        /// GetValue
        ///</summary>
        ///<param name="value"></param>
        ///<returns></returns>
        ///<exception cref="Exception"></exception>
        public virtual string GetValue(object value)
        {
            GridPanel panel = _Cell.GridPanel;

            if (value == null ||
                (panel.NullValue == NullValue.DBNull && value == DBNull.Value))
            {
                return ("");
            }

            return (Convert.ToString(value));
        }

        #endregion

        #region IGridCellEditControl Members

        #region Public properties

        #region CanInterrupt

        /// <summary>
        /// Gets whether the Grid can automatically interrupt the current
        /// NonModal cell edit operation when the mouse leaves the cell.
        /// 
        /// This is utilized in cases, such as the ButtonX editor, where
        /// multilevel controls may be involved.  In the ButtonX case this
        /// specifically enables the mouse to to leave the main cell window
        /// and not be interrupted by the grid until the current operation
        /// is complete by the editor itself.
        /// </summary>
        public bool CanInterrupt
        {
            get { return (true); }
        }

        #endregion

        #region CellEditMode

        /// <summary>
        /// Gets the cell editor activation mode. There are three modes
        /// available for choice, and their usage depends upon the desired
        /// editor presentation, as well as how well the given underlying
        /// control interacts with the grid control.
        /// </summary>
        public CellEditMode CellEditMode
        {
            // InPlace – Edit operations occur 'in place', always live, always
            // in a state of interaction with the user. The CheckBoxEx, and
            // ButtonX editors are examples of this. 

            get { return (CellEditMode.InPlace); }
        }

        #endregion

        #region EditorCell

        /// <summary>
        /// Gets or sets the editor's current associated grid cell.
        /// </summary>
        public GridCell EditorCell
        {
            get { return (_Cell); }
            set { _Cell = value; }
        }

        #endregion

        #region EditorCellBitmap

        ///<summary>
        /// Gets or sets the editor's associated cell bitmap (internal use only).
        ///</summary>
        public Bitmap EditorCellBitmap
        {
            get { return (_EditorCellBitmap); }

            set
            {
                if (_EditorCellBitmap != null)
                    _EditorCellBitmap.Dispose();

                _EditorCellBitmap = value;
            }
        }

        #endregion

        #region EditorFormattedValue

        ///<summary>
        /// Gets the formatted editor value.
        ///</summary>
        public virtual string EditorFormattedValue
        {
            get { return (Text); }
        }

        #endregion

        #region EditorPanel

        /// <summary>
        /// Gets or sets the editor's associated cell panel (internal use only).
        /// </summary>
        public EditorPanel EditorPanel
        {
            get { return (_EditorPanel); }
            set { _EditorPanel = value; }
        }

        #endregion

        #region EditorValue

        /// <summary>
        /// Gets or sets the editor's value.
        /// </summary>
        public virtual object EditorValue
        {
            get { return (Value); }
            set { Value = GetValue(value); }
        }

        #endregion

        #region EditorValueChanged

        /// <summary>
        /// Gets or sets whether the editor value has changed.
        /// </summary>
        public virtual bool EditorValueChanged
        {
            get { return (_ValueChanged); }

            set
            {
                if (_ValueChanged != value)
                {
                    _ValueChanged = value;

                    // If the value is being marked as changed, then mark
                    // the editor session as 'dirty' as a feedback to the
                    // user of the intentional (or unintentional) change.

                    if (value == true)
                        _Cell.SetEditorDirty(this);
                }
            }
        }

        #endregion

        #region EditorValueType

        ///<summary>
        /// Gets the editor's default value data type.
        ///</summary>
        public virtual Type EditorValueType
        {
            get { return (typeof(string)); }
        }

        #endregion

        #region StretchBehavior

        /// <summary>
        /// Gets the editor's desired 'stretch' behavior.
        /// 
        /// The StretchBehavior defines whether the editor wants to
        /// automatically have it's size stretched to fill the given cell.
        /// 
        /// Editors, such as ButtonX, want to fill the cell horizontally
        /// and vertically, whereas other editors, such as the Slider,
        /// may only want to stretch horizontally (or potentially not at all). 
        /// </summary>
        public virtual StretchBehavior StretchBehavior
        {
            get { return (_StretchBehavior); }
            set { _StretchBehavior = value; }
        }

        #endregion

        #region SuspendUpdate

        /// <summary>
        /// Gets or sets whether updates to the grid
        /// state are (or should be) suspended.
        /// </summary>
        public bool SuspendUpdate
        {
            get { return (_SuspendUpdate); }
            set { _SuspendUpdate = value; }
        }

        #endregion

        #region ValueChangeBehavior

        /// <summary>
        /// Gets the behavior needed when a value changes in the editor.
        /// 
        /// For instance, CheckBoxEx Value changes require only the cell
        /// to be redisplayed with the new state, whereas editors such as
        /// the TextBoxX editor, may result in a new layout when the text changes.
        /// </summary>
        public virtual ValueChangeBehavior ValueChangeBehavior
        {
            get { return (ValueChangeBehavior.InvalidateLayout); }
        }

        #endregion

        #endregion

        #region InitializeContext

        ///<summary>
        /// Called to inform the editor it needs
        /// to initialize the its context environment.
        ///</summary>
        ///<param name="cell">Associated Cell</param>
        ///<param name="style">Effective cell style</param>
        public virtual void InitializeContext(GridCell cell, CellVisualStyle style)
        {
            _Cell = cell;

            Value = GetValue(_Cell.Value);
            Enabled = (_Cell.IsReadOnly == false);

            _ValueChanged = false;
        }

        #endregion

        #region GetProposedSize

        ///<summary>
        /// Called to retrieve the editors proposed size for the given
        /// cell, using the provided effective style and size constraint.
        ///</summary>
        ///<param name="g">Graphics object</param>
        ///<param name="cell">Associated Cell</param>
        ///<param name="style">Effective style</param>
        ///<param name="constraintSize">Constraining size</param>
        ///<returns>Proposed Size</returns>
        public virtual Size GetProposedSize(Graphics g,
            GridCell cell, CellVisualStyle style, Size constraintSize)
        {
            return (GetPreferredSize(constraintSize));
        }

        #endregion

        #region Edit support

        #region BeginEdit

        /// <summary>
        /// Called when a Modal cell edit is about to be initiated.
        /// </summary>
        /// <param name="selectAll">Signifies whether to select all
        /// editable content upon invocation.</param>
        /// <returns>The editor can cancel the edit
        /// operation by returning 'true' from this call</returns>
        public virtual bool BeginEdit(bool selectAll)
        {
            return (false);
        }

        #endregion

        #region EndEdit

        /// <summary>
        /// Called when the edit operation is about to end.
        /// </summary>
        /// <returns>The editor can cancel the end operation
        /// by returning 'true' from this call.</returns>
        public virtual bool EndEdit()
        {
            return (false);
        }

        #endregion

        #region CancelEdit

        /// <summary>
        /// Called when the edit operation is being cancelled.
        /// </summary>
        /// <returns>The editor can cancel the cancel operation
        /// by returning 'true' from this call.</returns>
        public virtual bool CancelEdit()
        {
            return (false);
        }

        #endregion

        #endregion

        #region CellRender

        /// Called to initiate the actual rendering of the editor
        /// value into the grid cell.  In most cases this can be (and is)
        /// handled by a simple call to EditorCell.CellRender(this, graphics).
        /// 
        /// If additional rendering is required by the editor, then the editor
        /// can completely render the cell contents itself (and never even call
        /// Editor.CellRender) or optionally perform additional rendering before
        /// or after the default cell rendering.
        public virtual void CellRender(Graphics g)
        {
            _Cell.CellRender(this, g);
        }

        #endregion

        #region Keyboard support

        #region CellKeyDown

        ///<summary>
        /// Called when a KeyDown occurs (not utilized when the
        /// CellEditMode is Modal)
        ///</summary>
        public virtual void CellKeyDown(KeyEventArgs e)
        {
            // If the user types a digit (0 - 9) then increment
            // the state color by 1 at that state offset

            if (e.KeyData >= Keys.D0 && e.KeyData <= Keys.D9)
            {
                int index = e.KeyData - Keys.D0;
                string value = (string)EditorCell.Value;
                string[] s = value.Split(',');

                int k = Math.Max(index + 1, s.Length);

                int[] n = new int[k];

                for (int i = 0; i < k; i++)
                    n[i] = i < s.Length ? s[i][0] - '0' : 0;

                n[index] = (n[index] + 1) % 5;

                StringBuilder sb = new StringBuilder();

                for (int i = 0; i < k; i++)
                {
                    sb.Append(n[i]);
                    sb.Append(",");
                }

                sb.Length--;

                EditorCell.Value = sb.ToString();
            }
            else
            {
                OnKeyDown(e);
            }
        }

        #endregion

        #region WantsInputKey

        /// <summary>
        /// Called to determine if the editor
        /// wants to process and handle the given key.
        /// </summary>
        /// <param name="key">Key in question</param>
        /// <param name="gridWantsKey">true if editor wants to process the key</param>
        /// <returns></returns>
        public virtual bool WantsInputKey(Keys key, bool gridWantsKey)
        {
            // Lets handle all the digit keys

            if (key >= Keys.D0 && key <= Keys.D9)
                return (true);

            // And let the grid handle the rest

            return (gridWantsKey == false);
        }

        #endregion

        #endregion

        #region Mouse support

        #region OnCellMouseMove

        ///<summary>
        /// OnCellMouseMove
        ///</summary>
        ///<param name="e"></param>
        public virtual void OnCellMouseMove(MouseEventArgs e)
        {
            OnMouseMove(e);
        }

        #endregion

        #region OnCellMouseEnter

        ///<summary>
        /// OnCellMouseEnter
        ///</summary>
        ///<param name="e"></param>
        public virtual void OnCellMouseEnter(EventArgs e)
        {
            OnMouseEnter(e);
        }

        #endregion

        #region OnCellMouseLeave

        ///<summary>
        /// OnCellMouseLeave
        ///</summary>
        ///<param name="e"></param>
        public virtual void OnCellMouseLeave(EventArgs e)
        {
            OnMouseLeave(e);
        }

        #endregion

        #region OnCellMouseUp

        ///<summary>
        /// OnCellMouseUp
        ///</summary>
        ///<param name="e"></param>
        public virtual void OnCellMouseUp(MouseEventArgs e)
        {
            OnMouseUp(e);
        }

        #endregion

        #region OnCellMouseDown

        ///<summary>
        /// OnCellMouseDown
        ///</summary>
        ///<param name="e"></param>
        public virtual void OnCellMouseDown(MouseEventArgs e)
        {
            OnMouseDown(e);
        }

        #endregion

        #endregion

        #endregion
    }
}
